bn.c

```diff
@@ -273,6 +273,45 @@ void bignum_mul(struct bn* a, struct bn* b, struct bn* c)
 }
 
 
+void bignum_mul_alt(struct bn *a, struct bn *b, struct bn *c)
+{
+  require(a, "a is null");
+  require(b, "b is null");
+  require(c, "c is null");
+
+  bignum_init(c);
+  DTYPE_TMP tmp = 0;
+  DTYPE tmp_to_add = 0;
+
+  int usable_len = BN_ARRAY_SIZE;
+
+  /*  this section speads up algorithm by "cutting" len of bignum*/
+  for (int i = BN_ARRAY_SIZE - 1; i >= 0; --i)
+  {
+    if (a->array[i] != 0 || b->array[i] != 0)
+    {
+        usable_len = 2 * (i + 1);
+        break;
+    }
+  }
+
+  usable_len = usable_len > BN_ARRAY_SIZE ? BN_ARRAY_SIZE : usable_len;
+  //
+
+  for (int i = 0; i < usable_len; ++i)
```

Comment:
This part is brilliant. I love how you've reduced the allocation of two bignum's into none. That really cuts straight to the bone :+1: 

However I would like to avoid the scoped-declarations of variables inside the for-loop for stylistic reasons and to be a bit more portable against shitty/subset compilers. I will pull it as-is and then edit it myself, no worries.

Reply:
Declarations in for loops are provided by C standard, so if compiler does not support the standard it is not a compiler but shit. It is better style to use temporary variables like iterators in scopes. But if you want to keep same style it's ok for me.
The main improvement is reducing time complexity.

Reply:
The old multiplication is O(n^3). New is *only* O(n^2).